src/boot.lisp

```diff
@@ -409,66 +411,126 @@
     `(multiple-value-call (lambda ,gvars ,@setqs)
        ,@form)))
 
+;;; types utils
+(defun object-type-code (object) (oget object "dt_Name"))
+(defun set-object-type-code (object tag) (oset tag object "dt_Name"))
+
+;;; for to accurately definition  LIST and CONS forms
+;;; used:
+;;; (typep form
+;;;    (cons
+;;;      (cond ((true-cons-p form) .... code sensetive for cons)
+;;;            (t  ...))))
+;;;
+(defun true-cons-p (form)
+  (%js-try
+   (progn
+     (list-length form)
+     nil)
+   (catch (err)
+     t)))
+
+;;; pure list predicate: (true-list-p (cons 1 2)) => nil
+;;;                      (listp (cons 1 2)) => t
+(defun true-list-p (obj) (and (consp obj) (not (true-cons-p obj))))
 
+;;; types predicate's
 ;;; mop predicate
 (defun mop-object-p (obj)
     (and (consp obj)
-         (eq (oget obj "tagName") :mop-object)
+         (eql (object-type-code obj) :mop-object)
          (= (length obj) 5)))
 
-;;; js-object predicate
+;;; future upgrade
+(defun clos-object-p (object) (eql (object-type-code object) :clos_object))
+
 (defun js-object-p (obj)
   (if (or (sequencep obj)
           (numberp obj)
           (symbolp obj)
           (functionp obj)
+          (characterp obj)
           (packagep obj))
       nil
       t))
 
-;; Incorrect typecase, but used in NCONC.
-(defmacro typecase (x &rest clausules)
+;;; symbol-function from compiler macro
+(defun functionp (f) (functionp f))
+(defun js-null-p (obj) (js-null-p obj))
+(defun js-undefined-p (obj) (js-undefined-p obj))
+
+(defun true () t)
+(defun false () nil)
+(defun void () (values))
+
+
+;;; types family
+(defun %check-type-error (place value typespec string)
```

Comment:
I guess you tried already, but it would be really nice to put this under src/types.lisp and keep boot.lisp minimal.

I guess there are bootstrapping issues.  Maybe I can help with it though?

Reply:
Thank. I don't think there could be a problem with that.
But on occasion I will ask for help 🙂

Reply:
Apparently the entire `typecase family` should be left in the `boot`. Macro is used when compiling all the code below in the bundle. Also as `check-type`.

I don't see a simple solution. Does it exist?

Reply:
hm.. i see!  🤔

It is definitely quite heavy, and it'll grow more complex still.

What if we keep a simpler typecase (like the one it was there), with another name, only for bootstrapping reasons?

Reply:
To make  simple `typecase` and then override?

In this case, all the upper functions compiled before the `types.lisp` will work in simple code, through the `!typep`. 

Since they are in the core of the runtime, this will be somewhat slower than it could be.

Is the small size of the bootloader worth it that?

I'm not sure  that `C` tricks easy to implement at the environment JS. Is it really necessary?

Reply:
Another case with function `object-type-code`.

The function is used in several places where a complex type is generated : it is a def!struct, a hash-table, a clos objects.

A single function can be abandoned by using direct code in different files. But it is difficult to maintain the system.

Also may be look at the option with a macro instead of a function.

Reply:
> In this case, all the upper functions compiled before the types.lisp will work in simple code, through the !typep.

Ah, that's true. 

> Is the small size of the bootloader worth it that?

I am not worry about the size so much, but with keeping the bootstrapping as straightforward as possible.

But it might be fine!  I will have a more careful look and see if there is a way to improve that.

Reply:
Loading is now and so easy 🙂

More than.

So I refused from the PCL and put the `clos` after all the basic functions. Higher it can no longer be raised.

Reply:
I would be doing this as:

1. implemented `types` in minimal functionality
2. implemented `structures`.  This will require changes in the `clos` (highgly likely).
3. Implemented the `format`. This will require changes in `stream` and `clos` - new `base stream` format and methods. The functionality of the `print` will also be expanded.
4. `clos` also needs to be freshened up. There are performance bottlenecks, and something is wrong with the functionality.🙂

After that, we can assume that the version of the system is functionally complete, and ready for the **redesign of the bundle**. 

And in this case, the decision will be more logical.

Maybe so?

Reply:
Also these two functions for an `object-type-code` can be implemented as a `compiler macro`.
These are exclusively internal functions.